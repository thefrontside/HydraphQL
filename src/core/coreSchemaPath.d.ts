declare;
{
  const coreSchemaPath: string;
  export default coreSchemaPath;
}
