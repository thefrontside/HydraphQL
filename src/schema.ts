import { printSchemaWithDirectives } from "@graphql-tools/utils";
import { transformSchema } from "./transformSchema.js";
import { CoreSync } from "./core/core.js";

export const schema = printSchemaWithDirectives(transformSchema([CoreSync()]));
