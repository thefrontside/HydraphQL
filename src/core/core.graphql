directive @field(
  at: _DirectiveArgument_!
  default: _DirectiveArgument_
) on FIELD_DEFINITION
directive @discriminates(
  with: _DirectiveArgument_
  opaqueType: String
) on INTERFACE
directive @discriminationAlias(
  value: String!
  type: String!
) repeatable on INTERFACE
directive @implements(interface: String!) on OBJECT | INTERFACE
directive @resolve(at: _DirectiveArgument_, from: String) on FIELD_DEFINITION

scalar _DirectiveArgument_

interface Node {
  id: ID!
}

type Query {
  node(id: ID!): Node
  nodes(ids: [ID!]!): [Node]!
}
