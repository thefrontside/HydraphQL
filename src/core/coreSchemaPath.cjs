/* eslint-env node */
/** @type {string} */
exports.coreSchemaPath = require.resolve("@frontside/hydraphql/core.graphql");
