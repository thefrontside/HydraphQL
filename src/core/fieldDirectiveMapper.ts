import _ from "lodash";
import type { GraphQLFieldConfig } from "graphql";
import type { FieldResolver, ResolverContext } from "../types.js";
import { id } from "../helpers.js";
import { HYDRAPHQL_EXTENSION } from "src/constants.js";

export function fieldDirectiveMapper(
  fieldName: string,
  field: GraphQLFieldConfig<
    { id: string },
    ResolverContext,
    Record<string, unknown> | undefined
  >,
  directive: Record<string, unknown>,
) {
  if (
    "at" in directive &&
    typeof directive.at !== "string" &&
    (!Array.isArray(directive.at) ||
      directive.at.some((a) => typeof a !== "string"))
  ) {
    throw new Error(
      `The "at" argument of @field directive must be a string or an array of strings`,
    );
  }
  const resolve = (field.resolve ?? id) as FieldResolver<unknown>;

  const fieldResolver: FieldResolver = ({ entity }, args, context, info) => {
    const source =
      (_.get(
        entity,
        (directive.at as undefined | string | string[]) ?? fieldName,
      ) as unknown) ?? directive.default;
    return resolve(source, args, context, info);
  };

  field.extensions = {
    ...field.extensions,
    [HYDRAPHQL_EXTENSION]: {
      fieldResolver,
    },
  };
  field.resolve = async ({ id }, args, context, info) => {
    const { loader } = context;
    const entity = await loader.load(id);
    if (!entity) return null;
    return fieldResolver({ id, entity }, args, context, info);
  };
}
