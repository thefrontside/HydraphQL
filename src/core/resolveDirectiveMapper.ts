import _ from "lodash";
import { connectionFromArray, ConnectionArguments } from "graphql-relay";
import {
  GraphQLInputObjectType,
  type GraphQLFieldConfig,
  type GraphQLInterfaceType,
  GraphQLInt,
  GraphQLString,
} from "graphql";
import type { DirectiveMapperAPI, ResolverContext } from "../types.js";
import {
  createConnectionType,
  decodeId,
  encodeId,
  getNodeTypeForConnection,
  isConnectionType,
  isNamedListType,
  unboxNamedType,
} from "../helpers.js";

export function resolveDirectiveMapper(
  fieldName: string,
  field: GraphQLFieldConfig<
    { id: string },
    ResolverContext,
    Record<string, unknown> | undefined
  >,
  directive: Record<string, unknown>,
  api: DirectiveMapperAPI & { typeName: string },
) {
  if (
    "at" in directive &&
    typeof directive.at !== "string" &&
    (!Array.isArray(directive.at) ||
      directive.at.some((a) => typeof a !== "string"))
  ) {
    throw new Error(
      `The "at" argument of @field directive must be a string or an array of strings`,
    );
  }

  if (isConnectionType(field.type)) {
    if (directive.nodeType && typeof directive.nodeType === "string") {
      const nodeType = getNodeTypeForConnection(
        directive.nodeType,
        (name) => api.typeMap[name],
        (name, type) => (api.typeMap[name] = type),
      );

      if (nodeType) field.type = createConnectionType(nodeType, field.type);
    } else {
      field.type = createConnectionType(
        api.typeMap.Node as GraphQLInterfaceType,
        field.type,
      );
    }

    if (field.args && Object.keys(field.args).length > 0) {
      const argsType = new GraphQLInputObjectType({
        name: `${api.typeName}${fieldName[0].toUpperCase()}${fieldName.slice(
          1,
        )}Args`,
        fields: { ...field.args },
      });
      field.args = { args: { type: argsType } };
    }

    field.args = {
      ...field.args,
      first: { type: GraphQLInt },
      after: { type: GraphQLString },
      last: { type: GraphQLInt },
      before: { type: GraphQLString },
    };

    field.resolve = async ({ id }, args, { loader }) => {
      if (directive.at === "id") return { id };

      const node = await loader.load(id);

      const source =
        (directive.from as string | undefined) ?? decodeId(id).source;
      const typename = unboxNamedType(field.type).name;
      const ref: unknown = _.get(node, directive.at as string | string[]);

      if (directive.at) {
        if (!ref) {
          return null;
        } else if (
          !Array.isArray(ref) ||
          ref.some((r) => typeof r !== "string")
        ) {
          throw new Error(
            `The "at" argument of @resolve directive for "${fieldName}" field must be resolved to an array of strings`,
          );
        }
      }

      const ids = ((ref ?? []) as string[]).map((r) => ({
        id: encodeId({
          source,
          typename,
          query: {
            ref: r as string | undefined,
            args: (args as { args: Record<string, unknown> }).args,
          },
        }),
      }));

      return {
        ...connectionFromArray(ids, args as ConnectionArguments),
        count: ids.length,
      };
    };
  } else {
    field.resolve = async ({ id }, args, { loader }) => {
      if (directive.at === "id") return { id };

      const node = await loader.load(id);

      const source =
        (directive.from as string | undefined) ?? decodeId(id).source;
      const typename = unboxNamedType(field.type).name;
      const isListType = isNamedListType(field.type);
      const ref: unknown = _.get(node, directive.at as string | string[]);

      if (directive.at) {
        if (!ref) {
          return null;
        } else if (
          isListType &&
          (!Array.isArray(ref) || ref.some((r) => typeof r !== "string"))
        ) {
          throw new Error(
            `The "at" argument of @resolve directive for "${fieldName}" field must be resolved to an array of strings`,
          );
        } else if (!isListType && typeof ref !== "string") {
          throw new Error(
            `The "at" argument of @resolve directive for "${fieldName}" field must be resolved to a string, but got "${typeof ref}"`,
          );
        }
      }

      return isListType
        ? ((ref ?? []) as string[]).map((r) => ({
            id: encodeId({
              source,
              typename,
              query: { ref: r as string | undefined, args },
            }),
          }))
        : {
            id: encodeId({
              source,
              typename,
              query: { ref: ref as string | undefined, args },
            }),
          };
    };
  }
}
