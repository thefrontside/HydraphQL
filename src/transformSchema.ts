import { Kind } from "graphql";
import type { DocumentNode } from "graphql";
import { makeExecutableSchema } from "@graphql-tools/schema";
import { validateSchema } from "graphql";
import type { Module, Resolvers } from "graphql-modules";
import { mapDirectives } from "./mapDirectives.js";
import type { FieldDirectiveMapper, GraphQLModule } from "./types.js";
import { CoreSync } from "./index.js";

export function transformSchema(
  additionalModules: (GraphQLModule | Module)[] = [],
  { generateOpaqueTypes }: { generateOpaqueTypes?: boolean } = {},
) {
  const modules = [CoreSync(), ...additionalModules];
  const directiveMappers: Record<string, FieldDirectiveMapper> = {};
  const typeDefs: DocumentNode[] = modules.flatMap((m) => {
    const { module: gqlModule, mappers = {} } = "id" in m ? { module: m } : m;
    const documents = gqlModule.typeDefs;
    documents.forEach((document) => {
      document.definitions.forEach((definition) => {
        if (definition.kind !== Kind.DIRECTIVE_DEFINITION) return;
        const directiveName = definition.name.value;
        const mapper = mappers?.[directiveName];
        if (mapper) directiveMappers[directiveName] = mapper;
      });
    });
    return documents;
  });
  const resolvers: Resolvers = modules.flatMap((m) => {
    const { config } = "id" in m ? m : m.module;
    return config.resolvers ?? [];
  });

  const schema = mapDirectives(makeExecutableSchema({ typeDefs, resolvers }), {
    directiveMappers,
    generateOpaqueTypes,
  });

  const errors = validateSchema(schema);

  if (errors.length > 0) {
    throw new Error(errors.map((e) => e.message).join("\n"));
  }
  return schema;
}
