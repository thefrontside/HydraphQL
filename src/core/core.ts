import { loadFiles, loadFilesSync } from "@graphql-tools/load-files";
import { createModule } from "graphql-modules";
import type { TypeDefs } from "graphql-modules";
import { createDirectiveMapperProvider } from "../mapperProvider.js";
import type { ResolverContext } from "../types.js";
import { fieldDirectiveMapper } from "./fieldDirectiveMapper.js";
import { resolveDirectiveMapper } from "./resolveDirectiveMapper.js";
import { coreSchemaPath } from "./coreSchemaPath.cjs";

/** @public */
export const CoreSync = (typeDefs: TypeDefs = loadFilesSync(coreSchemaPath)) =>
  createModule({
    id: "core",
    typeDefs,
    resolvers: {
      Node: {
        id: async (
          { id }: { id: string },
          _: never,
          { loader }: ResolverContext,
        ): Promise<string | null> => {
          const node = await loader.load(id);
          if (!node) return null;
          return id;
        },
      },
      Query: {
        node: (_: unknown, { id }: { id: string }): { id: string } => ({ id }),
        nodes: (_: unknown, { ids }: { ids: string[] }): { id: string }[] =>
          ids.map((id) => ({ id })),
      },
    },
    providers: [
      createDirectiveMapperProvider("field", fieldDirectiveMapper),
      createDirectiveMapperProvider("resolve", resolveDirectiveMapper),
    ],
  });

/** @public */
export const Core = async () => CoreSync(await loadFiles(coreSchemaPath));
