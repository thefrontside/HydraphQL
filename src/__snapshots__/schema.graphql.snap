directive @discriminates(opaqueType: String, with: _DirectiveArgument_) on INTERFACE

directive @discriminationAlias(type: String!, value: String!) repeatable on INTERFACE

directive @field(at: _DirectiveArgument_!, default: _DirectiveArgument_) on FIELD_DEFINITION

directive @implements(interface: String!) on INTERFACE | OBJECT

directive @resolve(at: _DirectiveArgument_, from: String) on FIELD_DEFINITION

interface Node {
  id: ID!
}

type Query {
  node(id: ID!): Node
  nodes(ids: [ID!]!): [Node]!
}

scalar _DirectiveArgument_